rust/connlib/clients/shared/src/lib.rs

```diff
@@ -276,38 +155,72 @@ where
     }
 }
 
-fn upload_interval() -> Interval {
-    let duration = upload_interval_duration_from_env_or_default();
-    let mut interval = tokio::time::interval_at(Instant::now() + duration, duration);
-    interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
-
-    interval
-}
-
-/// Parses an interval from the _compile-time_ env variable `CONNLIB_LOG_UPLOAD_INTERVAL_SECS`.
+/// Connects to the portal and starts a tunnel.
 ///
-/// If not present or parsing as u64 fails, we fall back to a default interval of 5 minutes.
-fn upload_interval_duration_from_env_or_default() -> Duration {
-    const DEFAULT: Duration = Duration::from_secs(60 * 5);
-
-    let Some(interval) = option_env!("CONNLIB_LOG_UPLOAD_INTERVAL_SECS") else {
-        tracing::warn!(interval = ?DEFAULT, "Env variable `CONNLIB_LOG_UPLOAD_INTERVAL_SECS` was not set during compile-time, falling back to default");
-
-        return DEFAULT;
-    };
-
-    let interval = match interval.parse() {
-        Ok(i) => i,
+/// When this function exits, the tunnel failed unrecoverably and you need to call it again.
+async fn connect<CB>(
+    api_url: Url,
+    token: SecretString,
+    device_id: String,
+    device_name_override: Option<String>,
+    os_version_override: Option<String>,
+    callbacks: CB,
+    max_partition_time: Option<Duration>,
+) where
+    CB: Callbacks + 'static,
+{
+    match connect_inner(
+        api_url,
+        token,
+        device_id,
+        device_name_override,
+        os_version_override,
+        callbacks.clone(),
+        max_partition_time,
+    )
+    .await
+    {
+        Ok(never) => match never {},
         Err(e) => {
-            tracing::warn!(interval = ?DEFAULT, "Failed to parse `CONNLIB_LOG_UPLOAD_INTERVAL_SECS` as u64: {e}");
-            return DEFAULT;
+            tracing::error!("Tunnel failed: {e:#}");
+            let _ = callbacks.on_disconnect(None);
```

Comment:
With these changes `on_disconnect` is always called with `None` for control channel errors? I think some clients need visibility into the http errors

Reply:
Yeah? The reconnection happens within `PhoenixChannel`. I'll double check if any of the clients need it. Problem is that this is now an `anyhow::Error` which I can't pass on to the callbacks.

Reply:
I just checked the clients and all of them just log this error.

Reply:
The problem is that the behavior changes if we pass an error or we don't, if we do pass an error the token is cleared.

I figure this is not the best behavior for our current architecture but a quick and dirty solution is passing an empty string.

@jamilbk  is there really a point right now where we want to call `onDisconnect` and not clear the token? 

If we don't want to clear the token we can just handle it internally.

Reply:
> The problem is that the behavior changes if we pass an error or we don't, if we do pass an error the token is cleared.

Ah okay, I didn't know that but I've now found this functionality in the Android client.

We can also remove the usage of `anyhow` from the `Eventloop` for now and actually pass the error on, that is probably a cleaner change. It is still a behaviour change because we always have an error so we will always clear the token now.

Reply:
> @jamilbk is there really a point right now where we want to call onDisconnect and not clear the token?

IIRC, I believe we call `onDisconnect(null)` to signal to the app that its call to `Session.disconnect()` completed successfully. At least that's how the Android and Apple behave.

So `onDisconnect(error)` is called for an unexpected disconnect (which today is only the 401 right?) and `onDisconnect(null)` is called for an expected disconnect.

Maybe it more clear to just name these explicitly:

- `onTokenExpired()`
- `onDisconnectCompleted()`

And not have to pass an error string at all because we currently don't handle the error case, right?

If the tunnel crashes or something, the apps have cleanup logic (e.g. `deinit()`) that perform any app-side cleanup. And the error will be logged in connlib's logfile anyway.

Reply:
https://github.com/firezone/firezone/issues/3993

Reply:
> It is still a behaviour change because we always have an error so we will always clear the token now.

This will be annoying to users because anytime they quit the client or stop the tunnel (i.e. `Session.disconnect`) they'll have to sign in again.

Reply:
> > It is still a behaviour change because we always have an error so we will always clear the token now.
> 
> This will be annoying to users because anytime they quit the client or stop the tunnel (i.e. `Session.disconnect`) they'll have to sign in again.

I thougt that is the same anyway? How can I stop the tunnel without signing out?

Reply:
> IIRC, I believe we call `onDisconnect(null)` to signal to the app that its call to `Session.disconnect()` completed successfully

That seems pointless to me. How can disconnect not be successful? We just de-allocate memory and close file descriptors 🤷‍♂️

IMO, the clients should be able to assume that the tunnel is instantly gone as soon as disconnect returns.

Reply:
> which today is only the 401 right?

No, could be any permanent error like the TUN device being broken or non-retryable errors in the portal connection (anything 4xx I think).

Reply:
> I thougt that is the same anyway? How can I stop the tunnel without signing out?

Reboot with Firezone running, Select "Disconnect and Quit" from the menubar, toggle the VPN connection off in Settings, etc.

>  At least on Android, I have to sign in every time.

If you press "sign out" then we explicitly sign you out. But we need the token to stick around until it's either deemed invalid or cleared by the user because for Always On vpn to work on Android across reboots it needs this.

This was tested yesterday by a customer, so I'm surprised it's getting cleared for you -- how are you closing the tunnel?

Reply:
Right, that does work yes! I didn't even think of toggling it that way.

Reply:
> > which today is only the 401 right?
> 
> No, could be any permanent error like the TUN device being broken or non-retryable errors in the portal connection (anything 4xx I think).

Actually that is not true. Any "fatal" error in connlib currently just gets logged! That is probably something we should fix eventually but for now, I can replicate that behaviour here.